lib/multi_server_socket.dart

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:async/async.dart';
+
+import 'src/utils.dart';
+
+/// The error code for an error caused by a port already being in use.
+final _addressInUseErrno = _computeAddressInUseErrno();
+int _computeAddressInUseErrno() {
+  if (Platform.isWindows) return 10048;
+  if (Platform.isMacOS) return 48;
+  assert(Platform.isLinux);
+  return 98;
+}
+
+/// An implementation of `dart:io`'s [ServerSocket] that wraps multiple servers
+/// and forwards methods to all of them.
+///
+/// This is useful for listening on multiple network interfaces while still
+/// having a unified way of controlling the servers. In particular, it supports
+/// serving on both the IPv4 and IPv6 loopback addresses using
+/// [MultiServerSocket.loopback].
+class MultiServerSocket extends StreamView<Socket> implements ServerSocket {
+  /// The wrapped servers.
+  final Set<ServerSocket> _servers;
+
+  /// Returns the port that one of the wrapped servers is listening on.
+  ///
+  /// If the wrapped servers are listening on different ports, it's not defined
```

Comment:
Should this be a `List<int>` instead? same for `address`. It seems a bit weird to have these apis that have undefined behavior.

Or maybe they should just throw if there is more than one.

Reply:
`List<int>` doesn't match the `ServerSocket` API that we implement.

I think returning the first port/address is generally fine, because they will be ports/addresses that connect to the server even if they aren't *unique*.

Reply:
Ah ok, consider adding @override to make that more clear but sgtm